homeassistant/components/plugwise/config_flow.py

```diff
@@ -20,14 +31,77 @@
 from homeassistant.helpers.typing import DiscoveryInfoType
 
 from .const import (  # pylint:disable=unused-import
+    CONF_USB_PATH,
     DEFAULT_PORT,
     DEFAULT_SCAN_INTERVAL,
     DOMAIN,
+    FLOW_NET,
+    FLOW_TYPE,
+    FLOW_USB,
+    PW_TYPE,
+    SMILE,
+    STICK,
+    STRETCH,
     ZEROCONF_MAP,
 )
 
 _LOGGER = logging.getLogger(__name__)
 
+CONF_MANUAL_PATH = "Enter Manually"
+
+CONNECTION_SCHEMA = vol.Schema(
+    {
+        vol.Required(FLOW_TYPE, default=FLOW_NET): vol.In(
+            {FLOW_NET: f"Network: {SMILE} / {STRETCH}", FLOW_USB: "USB: Stick"}
+        ),
+    },
+)
+
+
+@callback
+def plugwise_stick_entries(hass):
+    """Return existing connections for Plugwise USB-stick domain."""
+    sticks = []
+    for entry in hass.config_entries.async_entries(DOMAIN):
+        if entry.data.get(PW_TYPE) == STICK:
+            sticks.append(entry.data.get(CONF_USB_PATH))
+    return sticks
+
+
+async def validate_usb_connection(self, device_path=None) -> Dict[str, str]:
+    """Test if device_path is a real Plugwise USB-Stick."""
+    errors = {}
+    # Avoid creating a 2nd connection to an already configured stick
+    if device_path in plugwise_stick_entries(self):
+        errors[CONF_BASE] = "already_configured"
+        return errors, None
+
+    stick = await self.async_add_executor_job(plugwise.stick, device_path)
+    try:
+        await self.async_add_executor_job(stick.connect)
+        await self.async_add_executor_job(stick.initialize_stick)
+        await self.async_add_executor_job(stick.disconnect)
+    except PortError:
+        errors[CONF_BASE] = "cannot_connect"
+    except StickInitError:
+        errors[CONF_BASE] = "stick_init"
+    except NetworkDown:
+        errors[CONF_BASE] = "network_down"
+    except TimeoutException:
+        errors[CONF_BASE] = "network_timeout"
+    return errors, stick
```

Comment:
Please avoid all the jumps into the executor by wrapping them up in one function.

Reply:
@bdraco, first a question: why should this be avoided?

And, combining them in one function is not practical: there are 4 functions, we call use 3 of them here, and in usb.py we use all four of them but in a different combination.
Also, this happens in config_flow.py and usb.py, the functions in these 2 files are only called during init and when removing the integration, how much harm can that cause?

Reply:
Switching from async to sync context is expensive. We should try to avoid  doing it multiple times when we can do one call to the executor.